pkg/server/apiHandlerv2.go

```diff
@@ -672,3 +677,308 @@ func (h *apiHandlerv2) uuid(w http.ResponseWriter, r *http.Request) {
+		})
+	}
+	for _, versionInfo := range quote.GetVersions() {
+		switch versionInfo.VersionType {
+		case attest.AttestVersionType_ATTEST_VERSION_TYPE_EVE:
+			template.EveVersion = versionInfo.Version
+		case attest.AttestVersionType_ATTEST_VERSION_TYPE_FIRMWARE:
+			template.FirmwareVersion = versionInfo.Version
+		}
+	}
+	return &template
+}
+
+func quoteValidate(manager driver.DeviceManager, u uuid.UUID, quote *attest.ZAttestQuote, resp *attest.ZAttestQuoteResp) error {
+	deviceOptions, err := getDeviceOptions(manager, u)
+	if err != nil {
+		return err
+	}
+	if len(deviceOptions.Nonce) == 0 {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_NONCE_MISMATCH
+		return fmt.Errorf("empty nonce for device in controller")
+	}
+	attestDataHash := sha256.Sum256(quote.GetAttestData())
+	deviceCertsBytes, err := manager.GetCerts(u)
+	if err != nil {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_NO_CERT_FOUND
+		return fmt.Errorf("cannot get device certs: %s", err)
+	}
+	var deviceCerts common.Zcerts
+	err = json.Unmarshal(deviceCertsBytes, &deviceCerts)
+	if err != nil {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_INVALID
+		return fmt.Errorf("cannot unmarshal device certs: %s", err)
+	}
+	var signingCert *certs.ZCert
+	for _, el := range deviceCerts.Certs {
+		if el.Type == certs.ZCertType_CERT_TYPE_DEVICE_RESTRICTED_SIGNING {
+			signingCert = el
+			break
+		}
+	}
+	if signingCert == nil {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_NO_CERT_FOUND
+		return fmt.Errorf("cannot get device signing cert")
+	}
+	cert, err := x509Pem.ParseCert(signingCert.GetCert())
+	if err != nil {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_NO_CERT_FOUND
+		return fmt.Errorf("cannot parse device cert: %s", err)
+	}
+	err = verifySignature(quote.GetSignature(), attestDataHash[:], cert)
+	if err != nil {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_INVALID
+		return fmt.Errorf("cannot get device certs: %s", err)
+	}
+	attestData, err := tpm2.DecodeAttestationData(quote.GetAttestData())
+	if err != nil {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_INVALID
+		return fmt.Errorf("cannot decode attestation data: %s", err)
+	}
+	if attestData.Type != tpm2.TagAttestQuote {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_INVALID
+		return fmt.Errorf("wrong type of attestation data")
+	}
+
+	if !bytes.Equal(attestData.ExtraData, []byte(deviceOptions.Nonce)) {
+		resp.Response = attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_NONCE_MISMATCH
+		return fmt.Errorf("nonce missmatch")
+	}
+
+	pcrMap := map[int][]byte{}
+	expectedDigestAlgo := attestData.AttestedQuoteInfo.PCRSelection.Hash
+	var tpmHashAlgo attest.TpmHashAlgo
+	var hashObj hash.Hash
+	switch expectedDigestAlgo {
+	case tpm2.AlgSHA1:
+		tpmHashAlgo = attest.TpmHashAlgo_TPM_HASH_ALGO_SHA1
+		hashObj = sha1.New()
```

Comment:
The EVE API assumes we have sha256 so I don't understand why you have sha1 support here.